core/src/main/scala/com/gu/tableversions/core/TableVersions.scala

```diff
@@ -13,42 +17,96 @@ trait TableVersions[F[_]] {
+    }
 
-  /** Get details about partition versions in a table. */
-  def currentVersion(table: TableName): F[TableVersion]
+  /**
+    * Get details about partition versions in a table.
+    */
+  def currentVersion(table: TableName)(implicit F: Sync[F]): F[TableVersion] =
+    // Derive current version of a table by folding over the history of changes
+    // until either the latest or version marked as 'current' is reached.
+    for {
+      ts <- tableState(table)
+      matchingUpdates = ts.updates.span(_.metadata.id != ts.currentVersion)
+      updatesForCurrentVersion = matchingUpdates._1 ++ matchingUpdates._2.take(1)
+      operations = updatesForCurrentVersion.flatMap(_.operations)
+    } yield
+      if (isSnapshotTable(operations))
+        latestSnapshotTableVersion(operations)
+      else
+        applyPartitionUpdates(PartitionedTableVersion(Map.empty))(operations)
+
+  /** Return the history of table updates, most recent first. */
+  def updates(table: TableName)(implicit F: Sync[F]): F[List[TableUpdateMetadata]] =
+    tableState(table).map(_.updates.map(_.metadata).reverse)
 
   /**
     * Update partition versions to the given versions.
-    * This performs no checking if data has been written to the associated paths but it will verify that these versions
-    * 1) haven't been committed before and 2) these are the next versions to be committed for each of the partitions.
     */
-  def commit(table: TableName, update: TableVersions.TableUpdate): F[CommitResult]
+  def commit(table: TableName, update: TableVersions.TableUpdate): F[Unit]
+
+  /**
+    * Set the current version of a table to refer to an existing version.
+    */
+  def setCurrentVersion(table: TableName, id: TableVersions.CommitId): F[Unit]
+
+  //
+  // Internal operations to be provided by implementations
+  //
+
+  /**
+    * Produce description of the current state of table.
+    */
+  protected def tableState(table: TableName): F[TableState]
+
+  /**
+    * Handle initialisation of a new table if it doesn't exist already.
+    */
+  protected def handleInit(table: TableName)(newTableState: => TableState): F[Unit]
 
 }
 
 object TableVersions {
 
-  /** A collection of updates to partitions to be applied and tracked as a single atomic change. */
-  final case class TableUpdate(
+  final case class TableUpdateMetadata(
+      id: CommitId,
       userId: UserId,
       message: UpdateMessage,
-      timestamp: Instant,
-      operations: List[TableOperation])
+      timestamp: Instant
+  )
 
-  final case class UpdateMessage(content: String) extends AnyVal
+  object TableUpdateMetadata {
+
+    def apply(userId: UserId, message: UpdateMessage, timestamp: Instant): TableUpdateMetadata =
+      TableUpdateMetadata(createId(), userId, message, timestamp)
+
+    private def createId(): CommitId = CommitId(UUID.randomUUID().toString)
+  }
+
+  final case class CommitId(id: String) extends AnyVal
```

Comment:
maybe commitId could have a UUID inside instead of a string ?
for consistency with the version class

Reply:
I had a think about this and I'm tempted to leave it as it is. It's less important here, as we're not using the string version of the ID to write file paths etc that will be parsed back later.